src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -417,13 +417,40 @@ private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAcce
                 throw new InvalidOperationException(SR.Format(SR.Error_CannotActivateAbstractOrInterface, type));
             }
 
-            if (!type.IsValueType)
+            ConstructorInfo[] constructors = type.GetConstructors(BindingFlags.Public | BindingFlags.Instance);
+
+            bool hasParameterlessPublicConstructor =
+                constructors.Any(ctor => ctor.GetParameters().Length == 0);
+
+            // if it's not a value type, and there are no constructors, nor a default constructor
+            if (!type.IsValueType && constructors.Length == 0 && !hasParameterlessPublicConstructor)
+            {
+                throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
+            }
+
+            if (constructors.Length > 1 && !hasParameterlessPublicConstructor)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                throw new InvalidOperationException(SR.Format(SR.Error_MultipleParameterisedConstructors, type));
+            }
+
+            if (constructors.Length == 1 && !hasParameterlessPublicConstructor)
+            {
+                ConstructorInfo constructor = constructors[0];
+                ParameterInfo[] parameters = constructor.GetParameters();
+
+                if (!CanBindToTheseConstructorParameters(parameters, out string invalidPropertyName))
                 {
-                    throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
+                    throw new InvalidOperationException(SR.Format(SR.Error_CannotBindToConstructorParameter, type, invalidPropertyName));
                 }
+
+                object?[] parameterValues = new object?[parameters.Length];
+
+                for (int index = 0; index < parameters.Length; index++)
+                {
+                    parameterValues[index] = BindParameter(parameters[index], config, options);
+                }
+
+                return constructor.Invoke(parameterValues);
```

Comment:
@layomia Can you take a look at this since S.T.J has to deal with the very similar problem of creating instances type using parameterized constructors. I would like the behavior of Microsoft.Extensions.Configuraiton and S.T.J to align as much is possible. I realize there are already many difference for various reasons. Default case-insensitivity is an example.

In [your doc describing the S.T.J behavior](https://github.com/layomia/dotnet_runtime/blob/11ff8a576aaccbf291ce8e3e9ba0bd0b04913d37/src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md#feature-behavior), I see the following:

> If a constructor parameter does not match with a property, InvalidOperationException will be thrown if deserialization is attempted.

Do you think that would make sense to do here? We could always relax the restriction later. Are there other behaviors or restrictions that would make sense to copy?

Reply:
Did you see this https://github.com/dotnet/runtime/pull/67258#discussion_r846333268?